# **WEEK 5 Assignment – Introduction to Deep Learning & Neural Networks with Keras**
Surya Prakash Karasala
 
Date: 03/22/2021

## **Importing Required Packages**

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [2]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

## Loading Data

Let's read the dataset into a pandas dataframe.

In [3]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

###**Let's check how many data points we have**.

In [4]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [5]:
concrete_data.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [6]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

### **Cleaning and Normalizing Data**

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [7]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [9]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [10]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [11]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [12]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Import Keras

#### Let's go ahead and import the Keras library

In [13]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [14]:
from keras.models import Sequential
from keras.layers import Dense

In [15]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [16]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [17]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

## Train and Test the Network

Let's call the function now to create our model.

In [18]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [19]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 0s - loss: 1621.4827
Epoch 2/50
23/23 - 0s - loss: 1605.1234
Epoch 3/50
23/23 - 0s - loss: 1588.9576
Epoch 4/50
23/23 - 0s - loss: 1573.1084
Epoch 5/50
23/23 - 0s - loss: 1557.2670
Epoch 6/50
23/23 - 0s - loss: 1541.2190
Epoch 7/50
23/23 - 0s - loss: 1525.0874
Epoch 8/50
23/23 - 0s - loss: 1508.3408
Epoch 9/50
23/23 - 0s - loss: 1490.9365
Epoch 10/50
23/23 - 0s - loss: 1472.9244
Epoch 11/50
23/23 - 0s - loss: 1454.0109
Epoch 12/50
23/23 - 0s - loss: 1434.1429
Epoch 13/50
23/23 - 0s - loss: 1413.5071
Epoch 14/50
23/23 - 0s - loss: 1392.2307
Epoch 15/50
23/23 - 0s - loss: 1369.9038
Epoch 16/50
23/23 - 0s - loss: 1347.0388
Epoch 17/50
23/23 - 0s - loss: 1323.2316
Epoch 18/50
23/23 - 0s - loss: 1298.5239
Epoch 19/50
23/23 - 0s - loss: 1273.5432
Epoch 20/50
23/23 - 0s - loss: 1247.1649
Epoch 21/50
23/23 - 0s - loss: 1220.5424
Epoch 22/50
23/23 - 0s - loss: 1192.2269
Epoch 23/50
23/23 - 0s - loss: 1163.3483
Epoch 24/50
23/23 - 0s - loss: 1133.1967
Epoch 25/50
23/23 - 0s - 

Next we need to evaluate the model on the test data.

In [20]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step - loss: 303.3272


303.32720947265625

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [21]:
from sklearn.metrics import mean_squared_error

In [22]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

303.3271928673564 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [23]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 146.55361938476562
MSE 2: 144.85629272460938
MSE 3: 88.15160369873047
MSE 4: 66.16210174560547
MSE 5: 59.309391021728516
MSE 6: 61.18647003173828
MSE 7: 55.101871490478516
MSE 8: 39.1931037902832
MSE 9: 42.88587951660156
MSE 10: 40.32435989379883
MSE 11: 40.06373977661133
MSE 12: 37.04703140258789
MSE 13: 44.89870834350586
MSE 14: 45.29454040527344
MSE 15: 37.2619743347168
MSE 16: 33.74528503417969
MSE 17: 37.17276382446289
MSE 18: 37.30353546142578
MSE 19: 35.234622955322266
MSE 20: 37.6098518371582
MSE 21: 35.54961013793945
MSE 22: 35.84113311767578
MSE 23: 30.323328018188477
MSE 24: 35.762184143066406
MSE 25: 36.938880920410156
MSE 26: 37.34746551513672
MSE 27: 33.95750045776367
MSE 28: 34.09174346923828
MSE 29: 40.38818359375
MSE 30: 37.73537063598633
MSE 31: 33.33854293823242
MSE 32: 33.33687973022461
MSE 33: 32.98442840576172
MSE 34: 36.74315643310547
MSE 35: 36.50661087036133
MSE 36: 42.18501663208008
MSE 37: 32.779876708984375
MSE 38: 37.66395950317383
MSE 39: 35.1326141